napari/_vispy/canvas.py

```diff
@@ -70,26 +200,398 @@ def _set_theme_change(self, theme: str):
+        """
+        self._process_mouse_event(mouse_release_callbacks, event)
+
+    def _on_mouse_wheel(self, event: MouseEvent) -> None:
+        """Called whenever mouse wheel activated in canvas.
+
+        Parameters
+        ----------
+        event : vispy.app.canvas.MouseEvent
+            The vispy mouse event that triggered this method.
+
+        Returns
+        -------
+        None
+        """
+        self._process_mouse_event(mouse_wheel_callbacks, event)
+
+    @property
+    def _canvas_corners_in_world(self) -> npt.NDArray:
+        """Location of the corners of canvas in world coordinates.
+
+        Returns
+        -------
+        corners : np.ndarray
+            Coordinates of top left and bottom right canvas pixel in the world.
+        """
+        # Find corners of canvas in world coordinates
+        top_left = self._map_canvas2world([0, 0])
+        bottom_right = self._map_canvas2world(self._scene_canvas.size)
+        return np.array([top_left, bottom_right])
+
+    def on_draw(self, event: DrawEvent) -> None:
+        """Called whenever the canvas is drawn.
+
+        This is triggered from vispy whenever new data is sent to the canvas or
+        the camera is moved and is connected in the `QtViewer`.
+
+        Parameters
+        ----------
+        event : vispy.app.canvas.DrawEvent
+            The draw event from the vispy canvas.
+
+        Returns
+        -------
+        None
+        """
+        # The canvas corners in full world coordinates (i.e. across all layers).
+        canvas_corners_world = self._canvas_corners_in_world
+        for layer in self.viewer.layers:
+            # The following condition should mostly be False. One case when it can
+            # be True is when a callback connected to self.viewer.dims.events.ndisplay
+            # is executed before layer._slice_input has been updated by another callback
+            # (e.g. when changing self.viewer.dims.ndisplay from 3 to 2).
+            displayed_sorted = sorted(layer._slice_input.displayed)
+            nd = len(displayed_sorted)
+            if nd > self.viewer.dims.ndisplay:
+                displayed_axes = displayed_sorted
+            else:
+                displayed_axes = self.viewer.dims.displayed[-nd:]
+            layer._update_draw(
+                scale_factor=1 / self.viewer.camera.zoom,
+                corner_pixels_displayed=canvas_corners_world[
+                    :, displayed_axes
+                ],
+                shape_threshold=self._scene_canvas.size,
+            )
+
+    def on_resize(self, event: ResizeEvent) -> None:
+        """Called whenever canvas is resized.
+
+        Parameters
+        ----------
+        event : vispy.app.canvas.ResizeEvent
+            The vispy event that triggered this method.
+
+        Returns
+        -------
+        None
+        """
+        self.viewer._canvas_size = tuple(self.size)
+
+    def add_layer_to_visual(self, napari_layer, vispy_layer) -> None:
```

Comment:
I find this name a bit confusing. I would just call this `add_layer`... But I guess it's partly because the visual was already generated? (see other comment about that).

Reply:
I still woulnd't call it `add_layer_to_visual`. You're not adding a layer to any visual, after all. Rather, you're adding the layer visual to the canvas?

Reply:
Changed with [02693da](https://github.com/napari/napari/pull/5432/commits/02693da662cc214afae2bfb4f817b9fe63db370c). The name is now add_layer_visual_mapping since we add a key, value pair to self.layer_to_visual, so I think the method name is informative like this.